python/ray/data/datasource/partitioning.py

```diff
@@ -0,0 +1,165 @@
+import posixpath
+from typing import Literal, Optional, List, Callable, Dict, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    import pyarrow
+
+from ray.data.datasource.file_based_datasource import _unwrap_protocol
+from ray.util.annotations import DeveloperAPI
+
+
+@DeveloperAPI
+class PathPartitioning:
+    """Partition scheme for path-based partition formats.
+
+    Path-based partition formats embed all partition keys and values directly in
+    their dataset file paths.
+    """
+
+    def __init__(
+        self,
+        style: Literal["hive", "dir"] = "hive",
+        base_dir: Optional[str] = None,
+        field_names: Optional[List[str]] = None,
+        filter_fn: Optional[Callable[[Dict[str, str]], bool]] = None,
+    ):
+        """Creates a new path-based dataset partition scheme.
+
+        Two path partition formats are currently supported - Hive and Directory.
+
+        For Hive Partitioning, all partition directories under the base directory
+        will be discovered based on "{key1}={value1}/{key2}={value2}" naming
+        conventions. Key/value pairs do not need to be presented in the same
+        order across all paths. Unpartitioned directory names may exist under the
+        base directory, and have their contents added to the dataset. If a
+        partition filter is defined, then each unpartitioned file will be passed
+        in as an empty dictionary.
+
+        For Directory Partitioning, all directories under the base directory will
```

Comment:
Hmm, is this a common partitioning style? If we only support the first (like Spark does) it may simplify the code.

Reply:
It is relatively common, but not as ubiquitous as hive-style partitioning. CloudTrail and Kinesis Firehose are two notable AWS service examples that use directory partitioning, so this is useful for parsing/filtering partitioned datasets that they produce.